Feat: autocomplete searchbox

agtenr opened the pull request:
## Type of change

- [ ] Bugfix
- [ ] Component improvement
- [x] New component
- [ ] New package
- [ ] Other

## Description
This PR contains an autocomplete search box:

- Extends the fluent ui search box
- Adds an onResolveSuggestion method for rendering autocomplete options
- Re-uses the accessible controls from Connect
- Add arrow down and tab controls

# Improvements:
This component currently supports all features for the connect implementation. Extra funcionality is needed for Search:
 - On focus --> for personal search history
 - Custom Renderer for adding icons/controls to suggestions

NickSevens commented on packages/dw-react-controls/src/components/AutocompleteSearchBox/AutocompleteSearchBox.types.ts:
```diff
@@ -0,0 +1,76 @@
+	 * The debounce time in milliseconds for the search box to start resolving the suggestions.
+	 * @default 0
+	 */
+	debounceTime?: number;
```
perhaps change this to `resolveDelay` to keep it consistent with https://developer.microsoft.com/en-us/fluentui#/controls/web/pickers#IBasePickerProps

NickSevens commented on packages/dw-react-controls/src/components/AutocompleteSearchBox/AutocompleteSearchBox.types.ts:
```diff
@@ -0,0 +1,76 @@
+	 * The aria description for the confirmation button.
+	 */
+	confirmationButtonAriaDescription?: string;
+}
```
i would also add `calloutProps` to be able to override the callout when necessary

NickSevens commented on packages/dw-react-controls/src/components/AutocompleteSearchBox/HighlightedSuggestion/HighlightedSuggestion.base.tsx:
```diff
@@ -0,0 +1,41 @@
+		const parts = React.useMemo(() => text.split(new RegExp(`^(${filter})`, "gi")), [text, filter]);
+
+		return (
+			<span>
```
why do we need the start/end spaces?

NickSevens commented on packages/dw-react-controls/src/components/AutocompleteSearchBox/HighlightedSuggestion/HighlightedSuggestion.base.tsx:
```diff
@@ -0,0 +1,41 @@
+						</span>
+					);
+				})}{" "}
+			</span>
```
`span` containing `span`s seems odd. Can this html be simplified?

NickSevens commented on packages/dw-react-controls/src/components/AutocompleteSearchBox/stories/AutocompleteSearchBox.stories.mdx:
```diff
@@ -0,0 +1,39 @@
+# AutocompleteSearchBox
+
+The autocomplete searchbox is a component that expands on the Fluent UI React Searchox with an autocomplete dropdown.
+The component is inspired and build upon the Microsoft example: [https://github.com/microsoft/autocomplete-search-box]()
```
```suggestion
The component is inspired and built upon the Microsoft example: [https://github.com/microsoft/autocomplete-search-box]()
```

NickSevens commented on packages/dw-react-controls/src/components/AutocompleteSearchBox/stories/AutocompleteSearchBox.stories.mdx:
```diff
@@ -0,0 +1,39 @@
+
+The autocomplete searchbox is a component that expands on the Fluent UI React Searchox with an autocomplete dropdown.
+The component is inspired and build upon the Microsoft example: [https://github.com/microsoft/autocomplete-search-box]()
+When typing the searchbox, autocomplete suggestions are dynamically fetched and displayed in a dropdown below the searchbox.
```
```suggestion
When typing in the searchbox, autocomplete suggestions are dynamically fetched and displayed in a dropdown below the searchbox.
```

NickSevens commented on packages/dw-react-controls/src/components/AutocompleteSearchBox/stories/AutocompleteSearchBox.stories.mdx:
```diff
@@ -0,0 +1,39 @@
+### Basic usage
+
+This example shows the basic usage of the AutocompleteSearchBox component.
+The only required property is an `onResolveSuggestions` funtion that return a promise with the suggestions.
```
```suggestion
The only required property is an `onResolveSuggestions` function that returns a promise with the suggestions.
```

NickSevens commented on packages/dw-react-controls/src/components/AutocompleteSearchBox/stories/AutocompleteSearchBox.stories.mdx:
```diff
@@ -0,0 +1,39 @@
+
+### Advance usage
+
+This example shows the advance usage of the AutocompleteSearchBox component.
```
```suggestion
This example shows the advanced usage of the AutocompleteSearchBox component.
```

NickSevens commented on packages/dw-react-controls/src/components/AutocompleteSearchBox/stories/AutocompleteSearchBox.stories.mdx:
```diff
@@ -0,0 +1,39 @@
+This example shows the advance usage of the AutocompleteSearchBox component.
+It is possible to control the value outside of the component by using the `value` and `onChange` properties.
+It also shows some examples of the debounce functionality and how to use the `onSearch` event.
+There is also an optoin to disable the suggestion highlighting.
```
```suggestion
There is also an option to disable the suggestion highlighting.
```

agtenr commented:
@NickSevens Thanks for the review, I updated the code:
 - Fixed typos
 - Renamed debounce property
 - Exposed calloutProps
 - Removed unnecessary spans/spaces in highlighted component

NickSevens reviewed (commented):
the `calloutProps` could probably be a bit more dynamic still (e.g., how would I now decide on the beak being visible?), but let's go ahead with the PR for now. :) This can be improved in the future if the need is there.
